In [47]:
import pandas as pd

In [48]:

url = "https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv"
df = pd.read_csv(url)




In [49]:
df

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                     170            3.0       159.0     3413.433759   
1                     130            5.0        97.0     3149.664934   
2                     170            NaN        78.0     3079.038997   
3                     220            4.0         NaN     2542.392402   
4                     210            1.0       140.0     3460.870990   
...                   ...            ...         ...             ...   
9699                  140            5.0       164.0     2981.107371   
9700                  180            NaN       154.0     2439.525729   
9701                  220            2.0       138.0     2583.471318   
9702                  230            4.0       177.0     2905.527390   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year  origin fuel_type         drivetrain  \
0             17.7        2003  Europe  Gasoline    All-wheel drive   
1             17.8        2007     USA  Gasoline  Front-wheel drive   
2             15.1        2018  Europe  Gasoline  Front-wheel drive   
3             20.2        2009     USA    Diesel    All-wheel drive   
4             14.4        2009  Europe  Gasoline    All-wheel drive   
...            ...         ...     ...       ...                ...   
9699          17.3        2013  Europe    Diesel  Front-wheel drive   
9700          15.0        2004     USA  Gasoline    All-wheel drive   
9701          15.1        2008     USA    Diesel    All-wheel drive   
9702          19.4        2011     USA    Diesel  Front-wheel drive   
9703          14.7        2005    Asia    Diesel    All-wheel drive   

      num_doors  fuel_efficiency_mpg  
0           0.0            13.231729  
1           0.0            13.688217  
2           0.0            14.246341  
3           2.0            16.912736  
4           2.0            12.488369  
...         ...                  ...  
9699        NaN            15.101802  
9700        0.0            17.962326  
9701       -1.0            17.186587  
9702        1.0            15.331551  
9703       -1.0            14.884467  

[9704 rows x 11 columns]

### Q1

In [50]:
pd.__version__

'2.3.1'

### Q1 Answer: Pandas version '2.3.1'

### Q2

In [51]:
df.shape

(9704, 11)

In [52]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9704 entries, 0 to 9703
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   engine_displacement  9704 non-null   int64  
 1   num_cylinders        9222 non-null   float64
 2   horsepower           8996 non-null   float64
 3   vehicle_weight       9704 non-null   float64
 4   acceleration         8774 non-null   float64
 5   model_year           9704 non-null   int64  
 6   origin               9704 non-null   object 
 7   fuel_type            9704 non-null   object 
 8   drivetrain           9704 non-null   object 
 9   num_doors            9202 non-null   float64
 10  fuel_efficiency_mpg  9704 non-null   float64
dtypes: float64(6), int64(2), object(3)
memory usage: 834.1+ KB


### Q2 Answer: There are 9704 records are in the dataset

### Q3 

In [53]:
df['fuel_type'].unique()

array(['Gasoline', 'Diesel'], dtype=object)

In [54]:
df['fuel_type'].nunique()

2

### Q3  Ansswer: The fuel types are 2


### Q4 

In [55]:
null_count = df.isnull().sum()
print(null_count)

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64


### Q4 Answer: 4 columns in the dataset have missing values

### Q5

In [56]:
# Filter for cars from Asia
asia_cars = df[df['origin'] == 'Asia']

# Find the maximum MPG
max_mpg_asia = asia_cars['fuel_efficiency_mpg'].max()
max_mpg_asia

np.float64(23.759122836520497)

### Q5 Answer: The maximum fuel efficiency of cars from Asia is 23.75

### Q6 

In [57]:
df['horsepower'].median()

np.float64(149.0)

In [58]:
df['horsepower'].mode()

0    152.0
Name: horsepower, dtype: float64

In [59]:
df['horsepower'] = df['horsepower'].fillna(152.0)
df['horsepower']

0       159.0
1        97.0
2        78.0
3       152.0
4       140.0
        ...  
9699    164.0
9700    154.0
9701    138.0
9702    177.0
9703    140.0
Name: horsepower, Length: 9704, dtype: float64

In [60]:
null_count = df.isnull().sum()
print(null_count)

engine_displacement      0
num_cylinders          482
horsepower               0
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64


In [61]:
df['horsepower'].median()

np.float64(152.0)

### Q6 Answer: Yes the Median value is changed to 152.0

### Q7


In [62]:
asia_cars = df[df['origin'] == 'Asia']


In [63]:
X_df = asia_cars[['vehicle_weight', 'model_year']].head(7)


In [64]:
import numpy as np

# Get the underlying NumPy array
X = X_df.to_numpy()
print("Matrix X (first 7 cars from Asia with 'Weight' and 'Model Year'):")
print(X)
# 4. Compute matrix-matrix multiplication between the transpose of X and X
XTX = X.T @ X
#Invert XTX
XTX_inv = np.linalg.inv(XTX)



Matrix X (first 7 cars from Asia with 'Weight' and 'Model Year'):
[[2714.21930965 2016.        ]
 [2783.86897424 2010.        ]
 [3582.68736772 2007.        ]
 [2231.8081416  2011.        ]
 [2659.43145076 2016.        ]
 [2844.22753389 2014.        ]
 [3761.99403819 2019.        ]]


In [65]:
# Create an array y
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

# Multiply the inverse of XTX with the transpose of X, and then by y
w = (XTX_inv @ X.T) @ y

# Get the sum of all the elements of the result
sum_of_w = w.sum()

print(f"\nThe final vector w is: {w}")
print(f"The sum of all the elements of w is: {sum_of_w}")


The final vector w is: [0.01386421 0.5049067 ]
The sum of all the elements of w is: 0.5187709081074016


### Q7 Answer: The sum of all the elements of w is: 0.5187709081074016